addons/account/models/account_reconcile_model.py

```diff
@@ -612,17 +612,19 @@ def _get_invoice_matching_query(self, st_lines_with_partner, excluded_ids):
                         (
                             (""" + self._get_select_communication_flag() + """)
                             OR
-                            (""" + self._get_select_payment_reference_flag() + """)
+                            (""" + self._get_select_payment_reference_flag() + r""")
                         )
                     )
                     OR
                     (
                         /* We also match statement lines without partners with amls
                         whose partner's name's parts (splitting on space) are all present
-                        within the payment_ref, in any order, with any characters between them. */
+                        within the payment_ref, in any order, with any characters between them.
+                        Escape the regexp characters before splitting the name. */
```

Comment:
@madprog thanks for the fix!

@oco-odoo @qdp-odoo @smetl  Gentlemen, I think there are multiple issues with this thing.

- The main issue is that you can't turn arbitrary text (data) into a regex (code) without escaping - just like you can't turn text into HTML markup, text into SQL, etc. This causes bugs *and* security problems, even with regexes. Hence @madprog's fix.
- If you look at a real-life contact list you will see that it comes with a lot of entries that aren't just "FirstName LastName", including the most common case: companies legal forms (sprl or s.p.r.l, gmbh, ltd, dmcc, etc..), expressed with or without punctuation. It seems the naive split should have provisions for that, such as ignoring punctuation, ignoring parts that are too short, etc.
- splitting on words is likely to cause small words in the partner name to create false positives in the payment ref. Considering the split, that could be mitigated by adding "word boundaries" (`\b` around the words being matched). Or perhaps we don't care about false positive here?
- considering that payment references may have charset restrictions, it's quite likely they may not have the full unicode version of the name - so both sides should probably go through `unaccent()` before attempting a match (when unaccent is available):
    ```
    =# select 'simsek' ~* unaccent('ŞİMŞEK');
     ?column? 
    ----------
     t
    (1 row)
    ```
- The `lower()` on the right hand side is not useful (because `~*` is case-insensitive), and could cause mismatches when the payment_ref has the same case as the contact, because some unicode codepoints have peculiar notions of lowercase (though `unaccent` would likely fix that too), e.g:
    ```
    =# select 'ŞİMŞEK' ~* lower('ŞİMŞEK');
     ?column? 
    ----------
     f
    (1 row)
    ```


And now I wonder: could we get rid of the escaping and only match alphanumeric sequences, replacing any other char in the substring by a wildcard? That could be simpler and also improve the chances of matching?

Reply:
I like the idea you suggest

Reply:
I agree with the idea of using wildcards as well.

Reply:
@odony The most recent version should address your remarks.